Wrappers/Python/cil/optimisation/functions/ApproximateGradientSumFunction.py

```diff
@@ -0,0 +1,237 @@
+        This sampler is called each time `gradient` is called and sets the internal `function_num` passed to the `approximate_gradient` function.  Default is `Sampler.random_with_replacement(len(functions))`. 
+
+    Note
+    -----
+    We ensure that the approximate gradient is of a similar order of magnitude to the full gradient calculation. For example, in the `SGFunction` we approximate the full gradient by :math:`n\nabla f_i` for an index :math:`i` given by the sampler. 
+    The multiplication by `math:`n` is a choice to more easily allow comparisons between stochastic and non-stochastic methods and between stochastic methods with varying numbers of subsets. 
+
+    Note
+    -----
+    Each time `gradient` is called the class keeps track of which functions have been used to calculate the gradient. This may be useful for debugging or plotting after using this function in an iterative algorithm:  
+        - `data_passes_indices` is a list of lists. Each time `gradient` is called a list is appended with with the indices of the functions have been used to calculate the gradient.  
+        - `data_passes` is a list. Each time `gradient` is called an entry is appended with  the proportion of the data used when calculating the approximate gradient  since the class was initialised (a full gradient calculation would be 1 full data pass). Warning: if your functions do not contain an equal `amount` of data, for example your data was not partitioned into equal batches, then you must first use the `set_data_partition_weights" function for this to be accurate.   
+
+
+
+    Note
+    ----
+    The :meth:`~ApproximateGradientSumFunction.gradient` returns the approximate gradient depending on an index provided by the  :code:`sampler` method. 
+
+    Example
+    -------
+    This class is an abstract base class, so we give an example using the SGFunction child class. 
+    
+    Consider the objective is to minimise: 
+
+    .. math:: \sum_{i=0}^{n-1} f_{i}(x) = \sum_{i=0}^{n-1}\|A_{i} x - b_{i}\|^{2}
+
+    >>> list_of_functions = [LeastSquares(Ai, b=bi)] for Ai,bi in zip(A_subsets, b_subsets))   
+    >>> f = ApproximateGradientSumFunction(list_of_functions) 
+
+    >>> list_of_functions = [LeastSquares(Ai, b=bi)] for Ai,bi in zip(A_subsets, b_subsets)) 
+    >>> sampler = Sampler.sequential(len(list_of_functions))
+    >>> f = SGFunction(list_of_functions, sampler=sampler)   
+    >>> f.full_gradient(x)
+    This will return :math:`\sum_{i=0}^{n-1} \nabla f_{i}(x)`
+    >>> f.gradient(x)
+    As per the approximate gradient implementation in the SGFunction this will return :math:`\nabla f_{0}`. The choice of the `0` index is because we chose a `sequential` sampler and this is the first time we called `gradient`. 
+    >>> f.gradient(x)
+    This will return :math:`\nabla f_{1}` because we chose  a `sequential` sampler and this is the second time we called `gradient`. 
```

Comment:
Maybe this could go in the tutorials section?

Reply:
I agree - we have an issue about this https://github.com/TomographicImaging/CIL/issues/1757. It is also in my todo list to create some demos. I think for the sake of this PR going in, we leave it as is at the moment